### PR TITLE
Add device flag

### LINKED ISSUE #73: Add device flag
avdmanager has a flag `--device` (or `-d`) that allows a user to specify what type of device to use. The default emulator is small, so this ticket is for creating an `androidEmulator.emulator.device` type for use in the gradle.

### PR DESCRIPTION
**Pull request by ap-ack**
Added support for the `device` field following the existing framework

### CODE REVIEW

File: android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java [hunk 1]
```diff
@@ -153,9 +153,14 @@ private static void createInstallEmulatorSystemImageTask(final Project project,
...
         final String emulatorName = emulatorConfiguration.getEmulatorName();
+        final String emulatorDevice = emulatorConfiguration.getDeviceType();
         createExecTask(project, emulatorConfiguration, CREATE_ANDROID_EMULATOR_TASK_NAME, exec -> {
                     exec.setExecutable(emulatorConfiguration.getAvdManager());
-                    exec.setArgs(l("create", "avd", "--name", emulatorName, "--package", emulatorConfiguration.getSystemImagePackageName(), "--force"));
+                    if (emulatorDevice == null) {
+                        exec.setArgs(l("create", "avd", "--name", emulatorName, "--package", emulatorConfiguration.getSystemImagePackageName(), "--force"));
+                    } else {
+                        exec.setArgs(l("create", "avd", "--name", emulatorName, "--package", emulatorConfiguration.getSystemImagePackageName(), "--device", emulatorDevice, "--force"));
+                    }
```
Comment (quittle) [inline android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java:163]: Could you instead create a list before and optionally extend if it `emulatorDevice` is not null? That way these two lists can't get out of sync with future changes.

File: README.md [hunk 2]
```diff
@@ -31,6 +31,7 @@ android {
 androidEmulator {
     emulator {
         name 'my_avd_emulator_name' // Defaults to be dynamically based on the configuration of the AVD
+        device 'pixel_xl' // or device '19'. For options, run avdmanager list device
```
Comment (quittle) [inline README.md:34]: Could you follow the convention of the rest of the arguments and explain what the default behavior is in addition to what you have now?

Comment (ap-ack) [reply README.md:34]: How about "Defaults to exclude the device flag, using avdmanager default. For options, run avdmanager list device"?

Comment (quittle) [reply README.md]: Yup, that works for me

**Comment by ap-ack** [comment]
Sounds good. I will get to this early next week.

**Comment by ap-ack** [comment]
Updated to include the tests. If you can pull my branch, does `validate_plugin` pass for you with these changes? I can't get it to pass at all on my machine, even before I made any changes. I will look into that too; might just be a local configuration problem.

File: android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java [hunk 3]
```diff
@@ -153,9 +153,15 @@ private static void createInstallEmulatorSystemImageTask(final Project project,
     private static void createCreateEmulatorTask(final Project project, final EmulatorConfiguration emulatorConfiguration) {
         final File avdRoot = emulatorConfiguration.getAvdRoot();
         final String emulatorName = emulatorConfiguration.getEmulatorName();
+        final String emulatorDevice = emulatorConfiguration.getDeviceType();
         createExecTask(project, emulatorConfiguration, CREATE_ANDROID_EMULATOR_TASK_NAME, exec -> {
                     exec.setExecutable(emulatorConfiguration.getAvdManager());
-                    exec.setArgs(l("create", "avd", "--name", emulatorName, "--package", emulatorConfiguration.getSystemImagePackageName(), "--force"));
+                    List<String> args = l("create", "avd", "--name", emulatorName, "--package", emulatorConfiguration.getSystemImagePackageName(), "--force");
+                    if (emulatorDevice != null) {
+                        args.add("--device");
```
Comment (quittle) [inline android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java:161]: I was able to reproduce the issue in CI and see the stacktrace by running `./gradlew -p example-android-project --stacktrace`

```
Caused by: java.lang.UnsupportedOperationException
        at com.quittle.androidemulator.AndroidEmulatorPlugin.lambda$createCreateEmulatorTask$8(AndroidEmulatorPlugin.java:161)
        at com.quittle.androidemulator.AndroidEmulatorPlugin.lambda$createExecTask$22(AndroidEmulatorPlugin.java:305)
        at org.gradle.api.internal.tasks.DefaultTaskContainer.create(DefaultTaskContainer.java:366)
        at com.quittle.androidemulator.AndroidEmulatorPlugin.createExecTask(AndroidEmulatorPlugin.java:302)
        at com.quittle.androidemulator.AndroidEmulatorPlugin.createCreateEmulatorTask(AndroidEmulatorPlugin.java:157)
        at com.quittle.androidemulator.AndroidEmulatorPlugin.lambda$apply$0(AndroidEmulatorPlugin.java:63)
        ...
```

I believe the issue is that my helper function `l()` uses `Arrays.asList()` which returns an immutable List so you can't just add an entry to it, hence the `UnsupportedOperationException`. I recommend updating the helper function to build and return a mutable list instead.

**Comment by quittle** [comment]
> I can't get it to pass at all on my machine, even before I made any changes. I will look into that too; might just be a local configuration problem.

What OS are you running on? It definitely works on Ubuntu and I believe it also works on Macs. Windows support is known to be buggy at this time.

**Comment by ap-ack** [comment]
I'm on a Mac.

File: android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java (hunk 3 above)
Comment (ap-ack) [reply android-emulator-plugin/src/main/java/com/quittle/androidemulator/AndroidEmulatorPlugin.java:161]: Nice catch. I made it an ArrayList.

**Comment by ap-ack** [comment]
Build passes now, should be good to go unless there's anything else you want me to add.

**Comment by quittle** [comment]
Congrats on getting the build to pass and thanks for trudging through that. Would you mind squashing your commits and giving it a good git commit description? I would prefer having just one commit for this change instead of 8 so it can be reverted if it breaks something. If you don't mind, I can rebase and update your commit description but I think the history will show me as the committer instead of you.

**Comment by quittle** [comment]
This has been released as version `0.2.1` of the plugin

**Pull request merged**